src/runtime_strings.js

```diff
@@ -9,11 +9,28 @@
 // Given a pointer 'ptr' to a null-terminated UTF8-encoded string in the given array that contains uint8 values, returns
 // a copy of that string as a Javascript String object.
 
+#if USE_PTHREADS && TEXTDECODER
+// UTF8Decoder.decode may not work with a view of a SharedArrayBuffer, see
+// https://github.com/whatwg/encoding/issues/172
+// To avoid that, we wrap around it and add a copy into a normal ArrayBuffer,
+// which can still be much faster than creating a string character by
+// character.
+function TextDecoderWrapper(encoding) {
+  var textDecoder = new TextDecoder(encoding);
+  this.decode = function(data) {
+    if (data.buffer instanceof SharedArrayBuffer) {
```

Comment:
Another note: TextDecoder.decode accepts any typed array, as well as ArrayBuffer / SharedArrayBuffer directly. This code will only correctly handle `Uint8Array`. Perhaps it's worth checking & throwing an explicit error if `data` is of any other type, or to support any type as well.

Reply:
I guess an assert can't hurt, but this code is not a general polyfill. It's just for the 3 places that we use TextDecoder internally, and those only use `Uint8Array`.

Reply:
Still, I imagine someone trying to use `.decode(buffer)` in the future and not knowing about this detail. Even conditional assert would be useful IMO.

Reply:
I see you added it now.